core/src/main/java/com/sequenceiq/cloudbreak/core/cluster/ClusterUpscaleService.java

```diff
@@ -94,15 +94,23 @@ public void uploadRecipesOnNewHosts(Long stackId, String hostGroupName) throws C
         recipeEngine.uploadUpscaleRecipes(stack, hostGroup, hostGroups);
     }
 
-    public void installServicesOnNewHosts(Long stackId, String hostGroupName) throws CloudbreakException {
+    public void installServicesOnNewHosts(Long stackId, String hostGroupName, Boolean repair, Boolean restartServices) throws CloudbreakException {
         Stack stack = stackService.getByIdWithClusterInTransaction(stackId);
-        LOGGER.debug("Start installing Ambari services");
+        LOGGER.debug("Start installing CM services");
         HostGroup hostGroup = Optional.ofNullable(hostGroupService.getByClusterIdAndNameWithRecipes(stack.getCluster().getId(), hostGroupName))
                 .orElseThrow(NotFoundException.notFound("hostgroup", hostGroupName));
         Set<InstanceMetaData> runningInstanceMetaDataSet = hostGroup.getInstanceGroup().getRunningInstanceMetaDataSet();
         recipeEngine.executePostAmbariStartRecipes(stack, hostGroup.getRecipes());
         ClusterApi connector = clusterApiConnectors.getConnector(stack);
         List<String> upscaledHosts = connector.upscaleCluster(hostGroup, runningInstanceMetaDataSet);
+        if (shouldRestartServices(repair, restartServices, stack)) {
+            try {
+                LOGGER.info("Trying to restart services");
+                connector.restartAll();
```

Comment:
How much time does it take to restart all the services? Does it make the repair a lot longer?

Reply:
around 3-4 minutes